src/renderer/rendered_window.rs

```diff
@@ -244,19 +260,21 @@ impl RenderedWindow {
         animating
     }
 
-    pub fn draw_surface(
-        &mut self,
-        canvas: &Canvas,
+    #[allow(clippy::type_complexity)]
+    fn split_lines(
```

Comment:
Maybe it would make sense to split this function into three different ones, since all three return values can be calculated completely independently? And also make it return iterators instead of vectors?